user-manager implementation

mpm11011 opened the pull request:
Looking for feedback on this. This is only a WIP and not ready to merge into rpcs3/master. 

- This does not contain a GUI yet; to change the user, you have to edit config.yml manually. 
- config.yml is probably not the place to put this... should it be in GuiConfigs/CurrentSettings.ini instead?
- A lot of Emu/Modules files now call Emu/System for the user id: will that negatively affect performance, readability, or design integrity? 
- I'm not happy with GetUsrId() implementation: any ideas on improving it?
- Considerations for savedata callbacks?

Megamouse commented:
why not call it GetUser and getUserId anyway? just wondering

mpm11011 commented:
Saw some other functions named like GetDir and GetCat in System.h. There's also TROPUSR. Figured I should follow suit from those.

Megamouse commented:
I don't mind it's just so short anyway it looks like a typo

tlm-2501 commented:
I would say config.yml is the correct place for storing this setting, and not CurrentSettings.ini. This is not a GUI-only setting. From my understanding, it will alter the way the emulator runs and how games search for saves and create them. Therefore, it is an emulation setting, and thus belongs in config.yml.

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.h:
```diff
@@ -74,6 +74,7 @@ private Q_SLOTS:
 	bool m_show_silver_trophies = true;
 	bool m_show_gold_trophies = true;
 	bool m_show_platinum_trophies = true;
+	std::string m_TROPHY_DIR;
```
no need for CAPS here anymore

mpm11011 replied:
Fixed in next commit.

VelocityRa commented on rpcs3/rpcs3qt/trophy_manager_dialog.h:
```diff
@@ -74,6 +74,7 @@ private Q_SLOTS:
 	bool m_show_silver_trophies = true;
 	bool m_show_gold_trophies = true;
 	bool m_show_platinum_trophies = true;
+	std::string m_TROPHY_DIR;
```
why capitals?

mpm11011 replied:
Probably some old code. Fixed it, see above.

VelocityRa commented on rpcs3/Emu/System.h:
```diff
@@ -247,6 +248,19 @@ class Emulator final
+	// TODO: I'm ashamed of this - cast AND type conversion. Must replace.
+	const u32 GetUsrId() const
+	{
+		return static_cast<u32>(std::stoul(m_usr));
```
There's no need for either the cast or the conversion, you can have it in a separate `u32` and only compute it once.

mpm11011 replied:
You're thinking store it in memory as a pointer, after its read from config.yml?

VelocityRa replied:
Yeah, simplest approach would be to have a separate `m_usr_id` and do the type conversion in the place where you assign to `m_usr` the val from the config.

Megamouse commented:
I agree that the current user has to be stored in config.yml.

Yet in the future one might add additional info like username and metadata.

Do you intend to add a GUI in this PR?

AniLeo commented:
Username is stored in dev_hdd0\home\USERID\localusername in a PS3, so it's better if you store it there.

mpm11011 commented:
@Megamouse Yes, but I am new to Qt so it may be a little while. Unless you think GUI should be in its own PR?

mpm11011 commented:
@AniLeo my idea for a GUI would be a 2-column table of User ID and the name stored in that file. Any other columns that should be there?

AniLeo commented:
I think that's fine. After user is selected, the rest of user data can be accessed via stuff like savedata manager, trophy manager, etc.

Megamouse commented:
A GUI for that is easily added. If you need help doing so you can consult me or others (preferably on Discord)

I think such a table should be fine for the initial purposes

mpm11011 commented:
Should it be its own window, like trophy manager, or should it be a tab under the settings dialog?

Megamouse commented:
I'd say use the menu entry in      Manage -> User Accounts    that already exists and make it open a simple dialog.

Hasster1 commented:
Yay! Finally, soon i will be able to play in terraria with my friend on split screen!!!

mpm11011 commented:
Basic GUI has been created, you can switch users while the main window is up, but you do need to re-initialize the emu in order for the new user to be stored in memory.

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+
+void user_manager_dialog::OnSort(int logicalIndex)
+{
+	if (logicalIndex >= 0)
```
return on < 1 instead

Megamouse commented on rpcs3/rpcs3qt/user_account.h:
```diff
@@ -0,0 +1,26 @@
+	std::string m_userDir;
+	std::string m_userName;
+
+};
```
newline

mpm11011 replied:
On the getters?

Megamouse replied:
no. end of file.
i didn't even notice the one you meant

mpm11011 replied:
Oh I see. I thought you meant you wanted the braces to be on separate lines for the getters.

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+
+	UpdateTable();
+
+	QByteArray geometry = m_gui_settings->GetValue(gui::sd_geometry).toByteArray();
```
why use savedata geometry ?

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+
+void user_manager_dialog::closeEvent(QCloseEvent * event)
+{
+	m_gui_settings->SetValue(gui::sd_geometry, saveGeometry());
```
same as before: why use savedata geometry ?

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+void user_manager_dialog::OnUserLogin()
+{
+	int idx = m_table->currentRow();
+	if (idx != -1)
```
return on -1 instead

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+void user_manager_dialog::OnUserRemove()
+{
+	int idx = m_table->currentRow();
+	if (idx != -1)
```
return on < 1 instead

flash-fire commented:
Q_Object should definitely be used in this case since you're using slots and connects. That macro tells QT's MOC (meta object compiler, what let's Qt do connects in first place with slots and stuff) to function. That'll create the autogenerated file that you can find.

From a cursory glance on my phone (lul), I can't tell why you're getting an error from that macro, but you should definitely be using it.

On top of my head, usually it's because you aren't including something that inherits QObject, but that's impossible given your includes. Otherwise, sometimes, the generated files DONT generate. So you got to cheese Qt a bit. It's simple. Remove qobject. Save. Readd it. Save again. It should pause a bit on saving. That's because it's forcefully rewriting the project's solution file to inform of the auto moc command.

If you still have issues, hit me up on discord. Current alias is chest

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+		userIdItem->setFlags(userIdItem->flags() & ~Qt::ItemIsEditable);
+
+		// Compare current config value with the one in this user (only 8 digits in userId)
+		if (m_emu_settings->GetSetting(emu_settings::SelectedUser)
```
take this out of the loop... and why this weird two liner

Megamouse replied:
also: you are using the exact same code a few lines below. just set both bold at the end of the loop

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+		userNameItem->setFlags(userNameItem->flags() & ~Qt::ItemIsEditable);
+
+		// Compare current config value with the one in this user (only 8 digits in userId)
+		if (m_emu_settings->GetSetting(emu_settings::SelectedUser)
```
see above

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+	for (UserAccount* user : m_user_list)
+	{
+		// For indicating logged-in user.
+		QFont boldFont;
```
take this out of the loop

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+	{
+		int originalIndex = m_table->item(row, 0)->data(Qt::UserRole).toInt();
+		UserAccount* originalEntry = m_user_list[originalIndex];
+		QString originalDirName = qstr(originalEntry->GetUserDir());
```
what's this for

Megamouse commented:
usually when you try building with the macro again after you removed it it will work. (i just tried)

Megamouse commented on rpcs3/rpcs3qt/user_account.h:
```diff
@@ -0,0 +1,26 @@
+class UserAccount
+{
+public:
+	explicit UserAccount(std::string userId);
```
const std::string& userId

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.h:
```diff
@@ -0,0 +1,53 @@
+{
+	// Q_OBJECT was here.
+public:
+	explicit user_manager_dialog(std::shared_ptr<gui_settings> gui_settings, std::shared_ptr<emu_settings> emu_settings, std::string dir = "", QWidget* parent = nullptr);
```
const std::string& dir

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,277 @@
+		m_table->verticalHeader()->width() + m_table->horizontalHeader()->length() + m_table->frameWidth() * 2,
+		m_table->horizontalHeader()->height() + m_table->verticalHeader()->length() + m_table->frameWidth() * 2);
+
+	QSize preferredSize = minimumSize().expandedTo(sizeHint() - m_table->sizeHint() + tableSize);
```
add `.expandedTo(size());` in order to fix unwanted size reset

mpm11011 replied:
Ah, that explains that.

mpm11011 commented:
Thanks for the very in depth review Mega! I'll fix and clean up all of this when I get out of work.

KD120 commented:
F {rsx::thread} class std::runtime_error thrown: Assertion Failed! Vulkan API call failed with unrecoverable error: Device lost (Driver crashed with unspecified error or stopped responding and recovered) (VK_ERROR_DEVICE_LOST)
(in file c:\projects\rpcs3\rpcs3\emu\rsx\vk\VKHelpers.h:906)

KD120 commented:
https://ci.appveyor.com/project/rpcs3/rpcs3/build/0.0.5-63365066/artifacts

KD120 commented:
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1830520/RPCS3.zip)

AniLeo commented:
Removed spam

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.h:
```diff
@@ -0,0 +1,53 @@
+
+#include <QDialog>
+#include <QTableWidget>
+#include <QIcon>
```
this is not used

mpm11011 commented:
Added a Create User modal dialog. Fixed some of the stuff @Megamouse mentioned above. 
- I don't know if how I did the Qt connect between the User Manager and the Create User dialog was proper, will fix if its not.
- Noticed the .gitignore files in /dev_hdd0/home/00000001... should these be removed? 

My next step is more negative testing and then a more thorough final review.

mpm11011 commented:
Oh gross... a backslash in "Utilities\StrUtil.h" How did that get there?

Megamouse commented:
Just from glancing at the code this does not look ideal. I'll review when i wake up tomorrow

mpm11011 commented:
Sounds good, thanks Mega.

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,320 @@
+	{
+		// "Entire string (^...$) must be between 3 and 16 characters
+		// and only consist of letters, numbers, underscores, and hyphens."
+		QRegExpValidator* validator = new QRegExpValidator(QRegExp("^[A-Za-z0-9_-]{3,16}$"));
```
this should not be a pointer

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,320 @@
+			QToolTip::showText(QCursor::pos(), tr("Username must have between 3 and 16 characters \nand only contain letters, numbers, underscores, and dashes."));
+			return;
+		}
+		user_manager_dialog::GenerateUser(textToValidate.toStdString());
```
I think writing out the class can be omitted, it takes longer to read and is unnecessary in my opinion

Megamouse commented on rpcs3/rpcs3qt/user_create_dialog.h:
```diff
@@ -0,0 +1,22 @@
+public:
+	explicit user_create_dialog(const std::string& dir = "", QWidget* parent = nullptr);
+	QString GetText();
+	QPushButton* GetOkButton();
```
replace GetOkButton with a Q_SIGNAL.
then use &user_create_dialog::<yourSignal> instead of &QAbstractButton::clicked in the user_manager_dialog

mpm11011 replied:
Yeah, I had a feeling it was odd to pass the button object around. So then I attach the signal to the ok button click in user_create_dialog?

Megamouse commented on rpcs3/rpcs3qt/user_create_dialog.cpp:
```diff
@@ -0,0 +1,42 @@
+QString user_create_dialog::GetText()
+{
+	return m_textBox->text();
+}
```
newline at end of file

Megamouse commented on rpcs3/rpcs3qt/user_create_dialog.cpp:
```diff
@@ -0,0 +1,42 @@
+	vbox_main->addLayout(hbox_buttons);
+	setLayout(vbox_main);
+
+	connect(m_cancelButton, &QAbstractButton::clicked, this, &user_create_dialog::close);
```
add
`connect(m_okButton, &QAbstractButton::clicked, this, &user_create_dialog::<your_signal>);`
(see other comment)

Megamouse commented on rpcs3/rpcs3qt/user_create_dialog.cpp:
```diff
@@ -0,0 +1,42 @@
+
+#include "user_create_dialog.h"
+
+user_create_dialog::user_create_dialog(const std::string& dir, QWidget* parent)
```
dir is unused and can be removed as param

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,320 @@
+		user_manager_dialog::UpdateTable();
+		createDialog->close();
+	});
+	createDialog->show();
```
the createDialog is not deleted on close.
This leads to hidden dialogs in background (hover over the user_manager_dialog's thumbnail in windows 10 and they will appear)

mpm11011 replied:
I did not know that! Thanks for pointing it out.

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,320 @@
+	menu->addSeparator();
+	menu->addAction(removeAct);
+	menu->addAction(loginAct);
+	menu->addAction(showDirAct);
```
it would probably be wise to add a "rename user" action

Megamouse commented:
changing anything breaks my compilation. my project files differ from yours after letting VS build.
I'll push them as commit soon and you can try if cherry-picking that works on your system still.

mpm11011 commented:
I wonder why that is. I use visual studio 2017, so to build the project I had to cherry pick one of zangetsu38's commits. But I made sure not to let that get included in this PR accidentally.

Megamouse commented:
i guess it's your handmade stuff. i will send you the commit later

mpm11011 commented:
RIP my handmade stuff. Thanks!

mpm11011 commented:
After adding the rename feature, I noticed that the /home/00000001/localusername file is tracked by git, so renaming that user makes a change in git. Should I be changing the .gitignore for that directory?

Megamouse commented:
it will not matter for regular users. so I'd just ignore the file locally

mpm11011 commented:
I added newlines in VS to each file, but they don't appear in Github... should I add another to each?

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -207,30 +210,72 @@ void user_manager_dialog::GenerateUser(const std::string& username)
+{
+	user_create_dialog* renameDialog = new user_create_dialog("New User", this);
+	renameDialog->setModal(true);
+	renameDialog->setAttribute(Qt::WA_DeleteOnClose, true);
```
you can just move these into the constructor, since they should be default.

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -207,30 +210,72 @@ void user_manager_dialog::GenerateUser(const std::string& username)
+	user_create_dialog* createDialog = new user_create_dialog("Rename User", this);
 	createDialog->setModal(true);
-	connect(createDialog->GetOkButton(), &QAbstractButton::clicked, this, [=]
+	createDialog->setAttribute(Qt::WA_DeleteOnClose, true);
```
you can just move these two lines into the constructor of user_create_dialog, since they should be default.

Megamouse commented on rpcs3/rpcs3qt/user_create_dialog.cpp:
```diff
@@ -2,10 +2,10 @@
 #include "user_create_dialog.h"
 
-user_create_dialog::user_create_dialog(const std::string& dir, QWidget* parent)
+user_create_dialog::user_create_dialog(const std::string& title, QWidget* parent)
```
use: `const QString& title` and move the tr() to the parts where you use it

Megamouse commented on rpcs3/rpcs3qt/user_create_dialog.cpp:
```diff
@@ -2,10 +2,10 @@
 	: QDialog(parent)
 {
-	setWindowTitle(tr("New User"));
+	setWindowTitle(tr(title.c_str()));
```
only use `setWindowTitle(title);`

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -99,6 +101,7 @@ void user_manager_dialog::Init(const std::string& dir)
 	connect(push_close, &QAbstractButton::clicked, this, &user_manager_dialog::close);
 	connect(push_remove_user, &QAbstractButton::clicked, this, &user_manager_dialog::OnUserRemove);
 	connect(push_create_user, &QAbstractButton::clicked, this, &user_manager_dialog::OnUserCreate);
+	connect(push_rename_user, &QAbstractButton::clicked, this, &user_manager_dialog::OnUserRename);
```
also add a connect that enables and disables the pushbuttons, depending on selected user

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -207,30 +210,72 @@ void user_manager_dialog::GenerateUser(const std::string& username)
+			const std::string dev_hdd0 = Emu.GetHddDir();
+			const std::string userId = m_user_list[idx_real]->GetUserId();
+			fs::remove_file(dev_hdd0 + "home/" + userId + "/localusername");
+			fs::write_file(dev_hdd0 + "home/" + userId + "/localusername", fs::create + fs::excl + fs::write, textToValidate.toStdString());
```
I think these can be done in just one step. with an additional fs::trunc.

also don't repeat these concatenations on the other parts and use a string instead

mpm11011 replied:
Can you explain how trunc would work here? Wouldn't it just cut the name down?

mpm11011 replied:
Disregard my last comment.

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.h:
```diff
@@ -29,11 +29,13 @@ private Q_SLOTS:
 	void Init(const std::string& dir);
 	void UpdateTable();
 	void GenerateUser(const std::string& username);
+	bool ValidateUsername(QString textToValidate);
```
const QString& textToValidate

Megamouse commented:
what the heck. it's all messed up

mpm11011 commented:
Probably need a member variable to track the currently logged in user, since changing it User Manager doesn't change it in Emu.

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,359 @@
+void user_manager_dialog::ShowContextMenu(const QPoint &pos)
+{
+	QPoint globalPos = m_table->mapToGlobal(pos);
+	QMenu* menu = new QMenu();
```
minor nitpick: you can move the top two lines bleow the idx check

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,359 @@
+		}
+		int idx_real = m_table->item(idx, 0)->data(Qt::UserRole).toInt();
+
+		bool enableButtons = m_user_list[idx_real]->GetUserId() != Emu.GetUsr();
```
yes. as you said: the usr Emu.GetUsr stays the same, so the pushbuttons will also stay consistently enabled/disabled

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,359 @@
+{
+	if (m_dir == "")
+	{
+		// fmt::format(%shome ... is harder to read than straight concatenation.
```
lol. funny remark. but you can just use concat. there's nothing wrong with it in this case.
same goes for the changes you made in the last commit. but that's really up to you

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,359 @@
+
+void user_manager_dialog::closeEvent(QCloseEvent * event)
+{
+	m_gui_settings->SetValue(gui::um_geometry, saveGeometry());
```
add: `QDialog::closeEvent(event);` at the end and please remove the space between * and event in the argument

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,359 @@
+
+void user_manager_dialog::OnUserRename()
+{
+	user_create_dialog* renameDialog = new user_create_dialog(tr("Rename User"), this);
```
The whole user_create_dialog class is pretty much useless.
You can just use a QInputDialog like pirhoman did in `settings_dialog::OnBackupCurrentConfig`.
The only downside to that would be that we cannot use the `setMaxLength(16)` if I remember correctly.

mpm11011 replied:
If my regular expression works as intended, that won't be a problem.

mpm11011 replied:
And yeah, that's what I get for not trying to find the best pre-defined class for this kind of work. Shame on me!

mpm11011 commented:
Small bug: Left clicking a non-logged in user, then right clicking it and hitting Login, it still leaves the bottom buttons for Rename and Delete enabled. I think I could disable them with a signal/slot, but I'm not sure how, short of making the buttons member variables...

Megamouse commented on rpcs3/rpcs3.vcxproj.filters:
```diff
@@ -605,6 +608,36 @@
+    <ClCompile Include="QTGeneratedFiles\Release\moc_user_create_dialog.cpp">
+      <Filter>Generated Files\Release</Filter>
+    </ClCompile>
+    <ClCompile Include="QTGeneratedFiles\Debug - LLVM\moc_user_create_dialog.cpp">
```
you forgot these 4 occurances of user_create_dialog

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,373 @@
+	m_selected_user = selectedUserId;
+	m_emu_settings->SetSetting(emu_settings::SelectedUser, m_selected_user);
+	m_emu_settings->SaveSettings();
+	UpdateTable();
```
you can just disable the buttons here, no need for new fancy connects

mpm11011 replied:
See that was my first thought, but I couldn't figure out how to retrieve the buttons from the dialog. I could make them member variables so I could access them from any function, but that felt like the "Get ok button" method, which I know you didn't like.

Megamouse replied:
of course you can either simply make them members or create another signal that gets emitted whenever you need to enable those buttons. eg:
``connect(this, &user_manager_dialog::OnUserLoginSuccess, enableButtons);``
and then use
``Q_EMIT OnUserLoginSuccess()``
on the appropriate place.

I don't really know what's the more appropriate way. members might be less confusing to maintain

mpm11011 replied:
I'll opt for easier maintenance. But I didn't know you could emit signals on demand like that, pretty cool!

mpm11011 replied:
Alright I lied - the Emit idea was definitely the easier to understand option.

Megamouse replied:
so is it all working now?

mpm11011 replied:
I think so!

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,375 @@
+	m_selected_user = m_emu_settings->GetSetting(emu_settings::SelectedUser);
+	UpdateTable();
+
+	QByteArray geometry = m_gui_settings->GetValue(gui::um_geometry).toByteArray();
```
I think you can just say: 
`restoreGeometry(m_gui_settings->GetValue(gui::um_geometry).toByteArray());`
since that will just return on invalid input.
I just changed that on the rest of those occurances. there would be no need for the if () anymore

mpm11011 commented:
Just rebased this branch on top of master.

mpm11011 commented:
Squashed so that I could rebase ontop of new changes more easily. Old state of the branch (before squash) is still around if I need to revert.

Hasster1 commented:
MERGE THIS PR!!!

Megamouse commented on rpcs3/Crypto/unself.cpp:
```diff
@@ -1396,7 +1397,7 @@ bool SELFDecrypter::GetKeyFromRap(u8* content_id, u8* npdrm_key)
 	// Try to find a matching RAP file under exdata folder.
 	const std::string ci_str = reinterpret_cast<const char*>(content_id);
-	const std::string rap_path = "/dev_hdd0/home/00000001/exdata/" + ci_str + ".rap";
+	const std::string rap_path = fmt::format("/dev_hdd0/home/%s/exdata/%s.rap", Emu.GetUsr(), ci_str);
```
According to KD-11's comment you should use:
`"/dev_hdd0/home/" + Emu.GetUsr() + "/exdata/" + ci_str + ".rap"`

Megamouse commented on rpcs3/Emu/Cell/Modules/cellSaveData.cpp:
```diff
@@ -1188,7 +1187,7 @@ s32 cellSaveDataUserFixedExport(ppu_thread& ppu, u32 userId, vm::cptr<char> dirN
 s32 cellSaveDataUserGetListItem(u32 userId, vm::cptr<char> dirName, vm::ptr<CellSaveDataDirStat> dir, vm::ptr<CellSaveDataSystemFileParam> sysFileParam, vm::ptr<u32> bind, vm::ptr<u32> sizeKB)
 {
-	cellSaveData.warning("cellSavaDataGetListItem(dirName=%s, dir=*0x%x, sysFileParam=*0x%x, bind=*0x%x, sizeKB=*0x%x, userID=*0x%x)", dirName, dir, sysFileParam, bind, sizeKB, userId);
+	cellSaveData.warning("cellSaveDataGetListItem(dirName=%s, dir=*0x%x, sysFileParam=*0x%x, bind=*0x%x, sizeKB=*0x%x, userID=*0x%x)", dirName, dir, sysFileParam, bind, sizeKB, userId);
```
the actual function name is `cellSaveDataUserGetListItem`

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNp.cpp:
```diff
@@ -65,8 +65,7 @@ s32 npDrmIsAvailable(vm::cptr<u8> k_licensee_addr, vm::cptr<char> drm_path)
 
-	// todo: profile for rap_dir_path
-	std::string rap_dir_path = "/dev_hdd0/home/00000001/exdata/";
+	std::string rap_dir_path = fmt::format("/dev_hdd0/home/%s/exdata/", Emu.GetUsr());
```
`"/dev_hdd0/home/" + Emu.GetUsr() + "/exdata/"`

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNp.cpp:
```diff
@@ -145,7 +144,7 @@ s32 sceNpDrmVerifyUpgradeLicense(vm::cptr<char> content_id)
 	}
 
-	if (!fs::is_file(vfs::get("/dev_hdd0/home/00000001/exdata/") + content_id.get_ptr() + ".rap"))
+	if (!fs::is_file(vfs::get(fmt::format("/dev_hdd0/home/%s/exdata/%s.rap", Emu.GetUsr(), content_id.get_ptr()))))
```
same as before use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNp.cpp:
```diff
@@ -164,7 +163,7 @@ s32 sceNpDrmVerifyUpgradeLicense2(vm::cptr<char> content_id)
 	}
 
-	if (!fs::is_file(vfs::get("/dev_hdd0/home/00000001/exdata/") + content_id.get_ptr() + ".rap"))
+	if (!fs::is_file(vfs::get(fmt::format("/dev_hdd0/home/%s/exdata/%s.rap", Emu.GetUsr(), content_id.get_ptr()))))
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -296,8 +296,7 @@ error_code sceNpTrophyRegisterContext(ppu_thread& ppu, u32 context, u32 handle,
 
-	// TODO: Get the path of the current user
-	std::string trophyPath = "/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name;
+	std::string trophyPath = fmt::format("/dev_hdd0/home/%s/trophy/%s", Emu.GetUsr(), ctxt->trp_name);
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -346,7 +345,7 @@ error_code sceNpTrophyGetRequiredDiskSpace(u32 context, u32 handle, vm::ptr<u64>
 	}
 
-	if (!fs::is_dir(vfs::get("/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name)))
+	if (!fs::is_dir(vfs::get(fmt::format("/dev_hdd0/home/%s/trophy/%s", Emu.GetUsr(), ctxt->trp_name))))
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -405,8 +404,7 @@ error_code sceNpTrophyGetGameInfo(u32 context, u32 handle, vm::ptr<SceNpTrophyGa
 
-	// TODO: Get the path of the current user
-	fs::file config(vfs::get("/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name + "/TROPCONF.SFM"));
+	fs::file config(vfs::get(fmt::format("/dev_hdd0/home/%s/trophy/%s/TROPCONF.SFM", Emu.GetUsr(), ctxt->trp_name)));
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -502,7 +500,7 @@ error_code sceNpTrophyUnlockTrophy(u32 context, u32 handle, s32 trophyId, vm::pt
 
 	ctxt->tropusr->UnlockTrophy(trophyId, 0, 0); // TODO
-	std::string trophyPath = "/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name + "/TROPUSR.DAT";
+	std::string trophyPath = fmt::format("/dev_hdd0/home/%s/trophy/%s/TROPUSR.DAT", Emu.GetUsr(), ctxt->trp_name);
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -524,7 +522,8 @@ error_code sceNpTrophyUnlockTrophy(u32 context, u32 handle, s32 trophyId, vm::pt
 		// Get icon for the notification.
-		std::string trophyIconPath = "/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name + "/TROP" + padding + std::to_string(trophyId) + ".PNG";
+		std::string paddedTrophyId = padding + std::to_string(trophyId);
+		std::string trophyIconPath = fmt::format("/dev_hdd0/home/%s/trophy/%s/TROP%s.PNG", Emu.GetUsr(), ctxt->trp_name, paddedTrophyId);
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -609,8 +608,7 @@ error_code sceNpTrophyGetTrophyInfo(u32 context, u32 handle, s32 trophyId, vm::p
 
-	// TODO: Get the path of the current user
-	fs::file config(vfs::get("/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name + "/TROPCONF.SFM"));
+	fs::file config(vfs::get(fmt::format("/dev_hdd0/home/%s/trophy/%s/TROPCONF.SFM", Emu.GetUsr(), ctxt->trp_name)));
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -750,7 +748,7 @@ error_code sceNpTrophyGetGameIcon(u32 context, u32 handle, vm::ptr<void> buffer,
 	}
 
-	fs::file icon_file(vfs::get("/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name + "/ICON0.PNG"));
+	fs::file icon_file(vfs::get(fmt::format("/dev_hdd0/home/%s/trophy/%s/ICON0.PNG", Emu.GetUsr(), ctxt->trp_name)));
```
use concat instead

Megamouse commented on rpcs3/Emu/Cell/Modules/sceNpTrophy.cpp:
```diff
@@ -803,7 +801,7 @@ error_code sceNpTrophyGetTrophyIcon(u32 context, u32 handle, s32 trophyId, vm::p
 	}
 
-	fs::file icon_file(vfs::get("/dev_hdd0/home/00000001/trophy/" + ctxt->trp_name + fmt::format("/TROP%03d.PNG", trophyId)));
+	fs::file icon_file(vfs::get(fmt::format("/dev_hdd0/home/%s/trophy/%s/TROP%03d.PNG", Emu.GetUsr(), ctxt->trp_name, trophyId)));
```
use concat instead

mpm11011 replied:
This one I'm gonna revert to the original way that uses fmt::format only for the trophyId. I think that format specifier is necessary for the PNG name, unless you know of a better method.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -1739,8 +1750,7 @@ void main_window::dropEvent(QDropEvent* event)
 
-			// TODO: use correct user ID once User Manager is implemented
-			if (!fs::copy_file(sstr(rap), fmt::format("%s/home/%s/exdata/%s", Emu.GetHddDir(), "00000001", rapname), false))
+			if (!fs::copy_file(sstr(rap), fmt::format("%s/home/%s/exdata/%s", Emu.GetHddDir(), Emu.GetUsr(), rapname), false))
```
use concat instead

Megamouse commented on rpcs3/rpcs3qt/trophy_manager_dialog.cpp:
```diff
@@ -56,6 +54,10 @@ trophy_manager_dialog::trophy_manager_dialog(std::shared_ptr<gui_settings> gui_s
 
+
+	// Get the currently selected user's trophy path.
+	m_trophy_dir = fmt::format("/dev_hdd0/home/%s/trophy/", Emu.GetUsr());
```
use concat instead

Megamouse commented on rpcs3/rpcs3qt/user_manager_dialog.cpp:
```diff
@@ -0,0 +1,371 @@
+	fs::create_dir(fmt::format("%s/exdata/", userDir));
+	fs::create_dir(fmt::format("%s/savedata/", userDir));
+	fs::create_dir(fmt::format("%s/trophy/", userDir));
+	fs::write_file(fmt::format("%s/localusername", userDir), fs::create + fs::excl + fs::write, username);
```
use concats for all of these instead

DeathDealer33 commented:
Anyway to manual add the users anymore? the yml im looking at is blank.

mkultimate commented:
the option is excellent, but it would be good in real time I think and add a manipulation using the joystick and try to avoid using the mouse I know that in the future the emulator will improve a lot and it would be a good option I'm just commenting as something alternative

mkultimate commented:
I append some examples I do not say that they are equal to these but it is a guide

![image](https://user-images.githubusercontent.com/16367065/42828190-c80f75fa-89ad-11e8-9a92-6cddf7c83386.png)

![image](https://user-images.githubusercontent.com/16367065/42828637-61decd34-89ae-11e8-9c90-b2733caaf6c4.png)

Nekotekina commented on rpcs3/Emu/Cell/Modules/cellSaveData.cpp:
```diff
@@ -904,7 +903,7 @@ s32 cellSaveDataListSave2(ppu_thread& ppu, u32 version, PSetList setList, PSetBu
 		version, setList, setBuf, funcList, funcStat, funcFile, container, userdata);
 
-	return savedata_op(ppu, SAVEDATA_OP_LIST_SAVE, version, vm::null, 1, setList, setBuf, funcList, vm::null, funcStat, funcFile, container, 2, userdata, 0, vm::null);
+	return savedata_op(ppu, SAVEDATA_OP_LIST_SAVE, version, vm::null, 1, setList, setBuf, funcList, vm::null, funcStat, funcFile, container, 2, userdata, Emu.GetUsrId(), vm::null);
```
Whenever savedata_op is called with zero userId, it's correct, it shouldn't be replaced.

Nekotekina commented on rpcs3/Emu/Cell/Modules/cellSaveData.cpp:
```diff
@@ -85,8 +85,8 @@ static NEVER_INLINE s32 savedata_op(ppu_thread& ppu, u32 operation, u32 version,
 
-	//TODO: get current user ID
-	// userId(0) = CELL_SYSUTIL_USERID_CURRENT
+	// TODO: now that we have Emu.GetUsrId(), do we need "userId ? userId : 1u"? Replace with error check?
```
GetUsrId should be here instead of 1u
